### ISSUE TITLE
Strategy for optional MyST syntax

### ISSUE THREAD

**Issue by hukkinj1** [hukkinj1/mdformat-myst#5]
The current state is that mdformat-myst only supports syntax that is enabled by default in myst-parser. User @ichard26 has voiced the need for optional syntax support (colon fence in particular). The purpose of this issue is to plan how support for optional syntax should be implemented.

I like that there is way to enable default syntax only so we shouldn't simply add optional syntax support the same way we do for non-optionals, as that would prevent installing "myst core" only. So that leaves us with a few options:

1. Add optional syntax support to this plugin but disable it by default. Add a CLI argument for enabling the optional syntax.
2. Make a `mdformat-myst-extensions` (or similar name) plugin that depends on `mdformat-myst` but adds optionals on top.
3. Very similar to option 2 but different naming. Make a `mdformat-myst-core` and move this plugin in its current state there. Then make this plugin require it and add optional syntax support.
4. Keep this plugin as "myst core" only and do not create a "mdformat full" plugin at all. Require separate plugin for each optional syntax that a user wishes to enable. We then end up having `mdformat-colon-fence`, `mdformat-amsmath` etc.

I think my favorite at least currently is option number 4, having the following strong points:
- keeps `mdformat-myst` simple and stable
- learning from the past: it's been useful how e.g. `tables` syntax is in its own plugin that can be collaboratively enabled by `mdformat-gfm` or `mdformat-myst` (whichever happens to run first)
- it's not yet clear if there's demand for any other optionals besides `colon-fence`
- transition from option 4 to any of the others is easy, if desired later

**Comment by hukkinj1** [comment]
@ichard26 sorry for a horribly long post, but in a nutshell, I propose that for `colon-fence` support we simply create `mdformat-colon-fence` plugin (or similar name), and for `MyST+colon_fence` support one would need to `pip install mdformat-myst mdformat-colon-fence`

I don't have the need or motivation to create that plugin right now myself, but will help and support you or anyone else willing to do so. [Here's](https://github.com/executablebooks/mdformat-plugin) a plugin template to get started btw.

**Comment by ichard26** [comment]
Sounds good, I'll look into it, totally new to this but it looks fun too!

**Comment by hukkinj1** [comment]
Some implementation pointers:

The colon fence extension for the Markdown parser lies in `mdit_py_plugins` PyPI package.

To enable that in the mdformat plugin one needs to do 

```python
from mdit_py_plugins.colon_fence import colon_fence_plugin

def update_mdit(mdit: MarkdownIt) -> None:
    mdit.use(colon_fence_plugin)
```

Then what is needed is the renderer function for `"colon_fence"` syntax. Probably should run the (already existing) renderer function of `"fence"` syntax, and use a regex or something to replace the leading and trailing `` ` `` characters with `:` chars.

**Comment by chrisjsewell** [comment]
yep for https://github.com/executablebooks/rst-to-myst/pull/18, will need colon_fence, deflist, and substitution (all from mdit_py_plugins)

**Comment by ichard26** [comment]
Soooooo, I totally forgot about this since psf/black ended up not using colon_fence, but I am still willing to implement this, although if you want it in a quick timeframe, don't count on me 😅

**Comment by chrisjsewell** [comment]
do you want to transfer this repo to executablebooks @hukkin

**Comment by hukkin** [comment]
Makes sense to me. Transferred!

**Comment by chrisjsewell** [comment]
yep you should have maintenance rights now 👍

**Comment by dylan-thomas832** [comment]
I would really like the ability to turn on/off MyST syntax, so I can easily use `mdformat-myst` on some libraries. Primarily, the colon fence feature is of interest to me. I know in #9 there is the idea of adding a `--myst-extensions` CLI argument, and I quite like that option, but I'm not sure what is needed to finish this and then start work on adding similar functionality for colon fences.

I just want to open the discussion back up and offer my help. What is the current block to progressing this forward?

Thanks for your work, `mdformat` is a great tool!

**Comment by LecrisUT** [comment]
I've made a [comment](https://github.com/executablebooks/mdformat/issues/234#issuecomment-1447613522) of another common strategy to control mdformat plugins.

**Comment by brunobeltran** [comment]
Probably a spicy take based on the comments above, but I would be willing to bet most users of a formatter realistically want a large set of (even non-default) plugins enabled by default. 

I can see many use cases where users of this plugin might want to use the formatter for various parts of a project (e.g., different books in one repo, each with different plugins installed), but the majority of the plugins being discussed here (IIUC) are _additive_ syntax. 

Consider:

- `repo_x`
   - `book_a/`
    - `book_b/`

If `book_a` has "colon_fence" enabled and `book_b` does not, then even if had colon fence support always-enabled with no flags to disable it---you could _still_ use a single venv (and a single `mdformat-myst` installation) to format both books. Why? Because `book_b` doesn't have the colon_fence plugin enabled, none of the documents of that book would _contain_ a colon-fence, so the colon-fence-specific rules for formatting would simply not be used! Meanwhile, our formatter would continue to happily receive `colon_fence` nodes when they _do_ appear (e.g., when formatting `book_a`).

It could be argued that this will pollute venvs unnecessarily---but honestly I doubt the extra tiny plugin deps would be burdensome to a typical user, and if they are for some users they can easily be segregated into a custom namespace specifically for formatting.

In short, I would propose that the mdformat-myst project specifically would strictly improve if we baked-in support for a large number of common (or at least a curated list of almost-always-used) non-default plugins, and actually hard-required them instead of trying to split this work out into a dozen different repos. 

Of course, happy to be told this is not the view of the current maintainers, and also happy to be told this type of support is out-of-scope for the current project (or that I'm missing something obvious).
